### PR TITLE
Phase08

### PR DESCRIPTION
**Pull request by rahaahmadi**

### CODE REVIEW

File: Phase08/Search/IOHandler.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            Console.WriteLine("Enter the input!");
+            return Console.ReadLine();
+        }
+        
+        public void PrintCommandResult(HashSet<string> commandResult)
+        {
+            if (!commandResult.Any())
+            {
+                Console.WriteLine("No do found!");
```
Comment (nedataghizadeh79) [inline Phase08/Search/IOHandler.cs]: "NO docs found!"

Comment (rahaahmadi) [reply Phase08/Search/IOHandler.cs]: =))

File: Phase08/Search/DocsFileReader.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+using System.IO;
+using System.Linq;
+using Search.Interface;
+using Search.Model;
+
+namespace Search
+{
+    public class DocsFileReader : IDocsFileReader
+    {
+        public HashSet<Doc> ReadContent(string path)
```
Comment (hoseinaghaei) [inline Phase08/Search/DocsFileReader.cs]: fekr konam phase ghabli in yechi dige return mikard
code ro bayad tori bezanin ke baray ijad ye feature dige code ghabli ro taghir nadin

File: Phase08/Search/FilterApplier.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            _invertedIndex = invertedIndex;
+        }
+        public HashSet<string> Filter(HashSet<string> plusWords, HashSet<string> minusWords, HashSet<string> noSignWords)
+        {
+            var plusCommandWordsDocs = GetSignDocs(plusWords);
+            var minusCommandWordsDocs = GetSignDocs(minusWords);
+            var noSignCommandWordsDocs = GetNoSignDocs(noSignWords);
+            var noSignWithoutMinus = new HashSet<string>(noSignCommandWordsDocs).Except(minusCommandWordsDocs).ToHashSet();
+            var result = !plusCommandWordsDocs.Any() ? noSignWithoutMinus : noSignWithoutMinus.Intersect(plusCommandWordsDocs).ToHashSet();
```
Comment (hoseinaghaei) [inline Phase08/Search/FilterApplier.cs]: in 2 3 khat bere ye tabe dige

File: Phase08/Search/InvertedIndex.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        }
+        
+        private void AddWordsDocs(Doc doc, IEnumerable<string> docWords)
+        {
+            foreach (var wordIter in docWords)
+            {
+                if (!_searchContext.Words.Any(w => w.Content == wordIter))
+                {
+                   AddNewWord(wordIter);
+                }
```
Comment (hoseinaghaei) [inline Phase08/Search/InvertedIndex.cs:51]: inja kalame ro ba doc khali add mikonin 
bad baz mirin peydash mikonin va doc ro behesh add mikonin
chera khob??????

File: Phase08/Search/Model/SearchContext.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public DbSet<Doc> Docs { get; set; }
+        public DbSet<Word> Words { get; set;}
+        public SearchContext(DbContextOptions<SearchContext> dbContextOptions) : base(dbContextOptions) { }
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            base.OnModelCreating(modelBuilder);
+            modelBuilder.Entity<Word>()
+            .HasMany(w => w.Docs)
+            .WithMany(d => d.Words).UsingEntity(x => x.ToTable("RelationTable"));
```
Comment (hoseinaghaei) [inline Phase08/Search/Model/SearchContext.cs]: esm table miani chera ine daghighan?

File: Phase08/Search/QueryProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Collections.Generic;
+using Search.Interface;
+
+namespace Search
+{
+    public class QueryProcessor : IQueryProcessor
+    {
+        public HashSet<string> PlusCommandWords { get; set; }
```
Comment (hoseinaghaei) [inline Phase08/Search/QueryProcessor.cs]: intori behtare :
public HashSet<string> PlusCommandWords { get; set; } = new();

File: Phase08/SearchTest/DocsFileReaderTest.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        private readonly IDocsFileReader _docsFileReader;
+
+        public DocsFileReaderTest()
+        {
+            _docsFileReader = new DocsFileReader();
+        }
+        
+        [Fact]
+        public void TestReadContentFromSingleFile()
```
Comment (hoseinaghaei) [inline Phase08/SearchTest/DocsFileReaderTest.cs]: esm tavabe test ro behtar bezarin
mesl kargah TDD

File: Phase08/SearchTest/FilterApplierTest.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        private readonly IInvertedIndex _invertedIndex;
+
+        public FilterApplierTest()
+        {
+            _invertedIndex = Substitute.For<IInvertedIndex>();
+        }
+      
+        [Theory]
+        [InlineData(new [] {"microsoft"},new [] {"xunit"},new [] {"hello"}, new string[] {})]
```
Comment (hoseinaghaei) [inline Phase08/SearchTest/FilterApplierTest.cs]: be jay inline data ke zesht mikone tabe ro az memberData estefade konin

File: Phase08/SearchTest/FullTextSearchTest.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            _fullTextSearch = new FullTextSearch(invertedIndex, docsFileReader,tokenizer, _queryProcessor, _filterApplier);
+        }
+
+        [Fact]
+        public void TestSearchQueryWithOneNoSignWord()
+        {
+            SetupMockedFilterApplier();
+            const string query = "hello";
+            _queryProcessor.MinusCommandWords.Returns(new HashSet<string>());
+            _queryProcessor.PlusCommandWords.Returns(new HashSet<string>());
```
Comment (hoseinaghaei) [inline Phase08/SearchTest/FullTextSearchTest.cs]: setup bere ye tabe dige

File: Phase08/SearchTest/FullTextSearchTest.cs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        }
+        
+        [Fact]
+        public void TestSearchQueryWithNoSignPlusMinusWords()
+        {
+            SetupMockedFilterApplier();
+            const string query = "hello +java -sample";
+            _queryProcessor.MinusCommandWords.Returns(new HashSet<string>(){"sample"});
+            _queryProcessor.PlusCommandWords.Returns(new HashSet<string>(){"java"});
+            _queryProcessor.NoSignCommandWords.Returns(new HashSet<string>(){"hello"});
```
Comment (hoseinaghaei) [inline Phase08/SearchTest/FullTextSearchTest.cs]: tabe dige

File: Phase08/SearchTest/InvertedIndexTest.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+namespace SearchTest
+{
+    public class InvertedIndexTest
+    {
+        private readonly ITokenizer _tokenizer;
+        private readonly IInvertedIndex _invertedIndex;
+
+        public InvertedIndexTest()
+        {
+            var contextOptions = new DbContextOptionsBuilder<SearchContext>()
```
Comment (hoseinaghaei) [inline Phase08/SearchTest/InvertedIndexTest.cs]: behtare bara inam ye factory bezarin mesl onvar

**Review by hoseinaghaei — COMMENTED** [review]
👌👌

File: Phase08/Search/InvertedIndex.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                    Name = docName,
+                    Content = docContent
+                };
+                AddWordsDocs(doc, docWords);
+            }
+        }
+        
+        public HashSet<string> GetWordDocs(string word)
+        {
+            var existingWord = _searchContext.Words.Include(w => w.Docs).FirstOrDefault(w => w.Content == word);
```
Comment (Parsa2820) [inline Phase08/Search/InvertedIndex.cs]: <p dir="rtl">
منطقی تره از SignleOrDefualt استفاده کنید
</p>

File: Phase08/Search/InvertedIndex.cs (hunk 4 above)
Comment (Parsa2820) [reply Phase08/Search/InvertedIndex.cs:51]: commandWords.Find(x => x.Content == wordIter)?.Docs.Add(doc) ?? AddNewWord(wordIter);

File: Phase08/Search/Model/SearchContext.cs [hunk 13]
```diff
@@ -0,0 +1,18 @@
+﻿using Microsoft.EntityFrameworkCore;
+
+namespace Search.Model
+{
+    public class SearchContext : DbContext
+    {
+        public DbSet<Doc> Docs { get; set; }
```
Comment (Parsa2820) [inline Phase08/Search/Model/SearchContext.cs:7]: <p dir="rtl">
indentation و enter و این چیزا رو رعایت کنید 
</p>

File: Phase08/SearchTest/FullTextSearchTest.cs (hunk 9 above)
Comment (rahaahmadi) [reply Phase08/SearchTest/FullTextSearchTest.cs]: bara har test setup fargh mikone, bara har kodom joda tabe bezanim?

**Review by Parsa2820 — APPROVED** [review]

**Review by hoseinaghaei — APPROVED** [review]

**Pull request merged**